### DISCUSSION TITLE
[Plugin] Rue progress overview (WIP)

### DISCUSSION THREAD

**Discussion by Bia10** [Kaioru/Edelstein#83] [Ideas]
This discussion tracks the progress on state of various undeveloped features/enhancements or present bugs as I slowly review the entire plugin.

My main perspective on the requirements presented is based on idea that Rue should provide 2 main points:

1.  Easy of use for E2E testing, as currently were still lack proper fully automated means of E2E testing **yet**, Rue should be able to offer partial solution for some aspects of that and offer easy way to attain any configuration of client states easily. Give a way to emulate/simulate/bypass certain events as to test them without having to go thru everything manually ingame which may be even impossible for lets say GM events or complex party quests.

2. Show what is already doable with the current state of affairs of neatly written codebase @Kaioru provided us with which was written and **rewritten** many times to take things like ease of extendibility and modifiability into account. As such offers a platform to test out protocol or game/login server code without actually having to modify the original code at all. What I personally consider an attractive aspect of modularity/isolability being part of the design.

### Generic issues

Lack of more descriptive logging of init and mainly start of plugins.

- [x] Logging of stage information. Done in d8656a7
- [ ] Logging of pipelines information.

CommandManager

- [ ] Add to `CommandManager` some kind of memory which would make a history of commands and would allow to move backwards/forwards in processing that history to undo/redo n command "steps".

Conversations

- [ ] Implement `ConversationTextFormatter` should generalize and unify approaches for generating text for conversation speakers. This implies using the client text formatting syntax to achieve things like positioning(left, right, middle, top, bottom), whitespace management(toBlock, toParagraph, insertEmptyLine(int count)), character formatting(coloring, italics, bold), hyperlinking(items, skills, characterName, mobs, quests), lists formatting(create, remove, fromText, reorder), progress reporting, pagination(offer ability to make book of pages... determine text length on each page, split into pages)
- [ ] To achieve complete `ToChat` option requires to establish a `ConversationChatContext`.

Implement better textual search handling

- [x] Investigate more performant implementation (Trie, Elasticsearch). Done in 655a949
- [ ] Investigate more friendly input processing -> spelling correction 

Record data storage, duplication, contextualization

- [ ]  Generalize Stats, Skill, Equip commands to handle entire recordSnapshot(s), something like base `IRecordSnapshot `-> `IStatRecordSnapshot`, `ISkillRecordSnapshot`, `IEquipRecordSnapshot` and implement on these `SetTo(IRecordSnapshot snapshot)` this gets you to wrappers/extensions like `SetToJob()`, `SetToDefault()`, `SetToQuestState()`, ultimately combining the snapshot interfaces into a `ICharacterBuildSnapshot`

### Login

- [x] Implement FlippedUserName handling done in #80
- [x] Implement AutoRegister feature done in #80
- [x] Implement AutoLogin feature done in #80
- [ ] Implement AutoWorldSelect feature
- [ ] Implement AutoCharacterCreate feature
- [ ] Implement AutoCharacterDelete feature
- [ ] Implement AutoCharacterSelect feature 
- [ ] Implement AutoIngame feature, the culmination of previous work -> get ingame as efficiently and smoothly as possible

### Game

Help Command

- [ ] Displayed commands only describe themselves, there is no reference to existence of args and their use at all.
- [ ] Add feature to call Help only on 1 or n commands.
- [ ] Additional information on the command, display its call syntax (the order of args, their possible values or ranges), display example use case(s)

Skill Command

- [ ] Add feature to remove skill(s).
- [ ] `ResetAll()` method merely sets all skill records to level 0, this doesn't imply the meaning of resetting to a default or previous state. This should be called `SetAllToZero()` as its common for an alt of a main chair to inherit some of skill levels in various blessings, for such char the default is not all records at 0 at all.

Stat Command

- [ ] The args are entirely unchecked will process any input as such can create char states which wont even pass character selection.
- [ ] Adding of sp actually doesn't discriminate between different categories of skill record collections so it gives sp to every job level progression.
- [ ] Implement pet related stat handling.

Debug Command

- [ ] Implement other missing records.
- [ ] Find way to format the text printed into chat window.
- [ ] `ToChat `args should also implement selection menu and answer in chat window.
- [ ] Debug other field entities.
- [ ] Allow for monitoring of objects of interest. Something along side observables of objects publishing events on trigger.
- [ ] The bold and blue formatting are not properly escaped and cause records to be also bold/blue which I find useful thus I consider this a feature until this behavior becomes issue.

**Comment by Kaioru** [comment]
Thanks for the large feedback on Rue! I'll send my thoughts here whenever I am ready.
Here are my initial thoughts so far:

> Easy of use for E2E testing, as currently were still lack proper fully automated means of E2E testing yet, Rue should be able to offer partial solution for some aspects of that and offer easy way to attain any configuration of client states easily. Give a way to emulate/simulate/bypass certain events as to test them without having to go thru everything manually in game which may be even impossible for lets say GM events or complex party quests.

Currently a large portion of the code is decoupled and abstracted heavily. Automated testing is possible (if not already possible) to achieve. However, I would suggest we wait on this because the main codebase is still constantly changing. Though, we can start looking into automating stuff that's already pretty much solidified.

>  Lack of more descriptive logging of init and mainly start of plugins

Yes I agree, I've added stage information during Init, Start, and Stop in d8656a7d43e0ff842844771eac4e287f45db826f
As for pipeline logging, if we were to only log plugin registrations, this would require a fundamental change in how pipelines and contexts are handled.

Otherwise, we could also simply log every pipeline registration; though the log level for that would have to be debug as there are a lot of pipeline registrations on startup.

>  Implement better textual search handling currently its either slow or lacking in effort put into working with the input, often returns nothing even when given half of the skill name.

This is not clearly stated anywhere in the codebase currently but the search command does allow for globbing. For example: !item "\*sword\*" would return with results with the word sword in it. See: https://github.com/dazinator/DotNet.Glob#patterns

As for efficiency and speed, yep. The currently implementation runs on a O(N) time complexity for each search, so for templates with large number of records such as Items would iterate over 13k strings every single time. There are a few alternatives I can think of:

1. Specially store the strings in a different structure (maybe a Trie) just for command searches.
2. Use a proper search engine like elasticsearch, though I feel this could add a lot of bloat and setup overhead for just a search command.

> Generalize Stats, Skill, Equip commands to handle entire recordSnapshot(s)

I'm not too quite sure what this one means.
